phases/19-capstone-projects/59-vit-transformer/code/test_main.py

```
@@ -0,0 +1,94 @@
+        self.assertIsNotNone(grad)
+        self.assertGreater(grad.norm().item(), 0.0)
+
+
+class TestEncoderConfig(unittest.TestCase):
+    def test_head_dim_divides(self) -> None:
+        cfg = ViTConfig(image_size=32, patch_size=16, hidden=64, depth=1, heads=4)
+        self.assertEqual(cfg.head_dim, 16)
+
+    def test_invalid_head_dim_rejected(self) -> None:
+        with self.assertRaises(ValueError):
+            ViTConfig(image_size=32, patch_size=16, hidden=65, depth=1, heads=4).head_dim
```

Ruff (0.15.14)

[warning] Found useless attribute access. Either assign it to a variable or remove it.

(B018)